numaresourcesscheduler: respond to owned objects` modifications

Tal-or opened the pull request:
Prior to this change, every external/manual tamper to the generated objects wouldn't be detected by the controller, hence it keeps the objects in an undesired state.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

Tal-or committed:
numaresourcesscheduler: respond to owned objects` modifications

Prior to this change, every external/manual tamper to the generated objects wouldn't be detected by the controller, hence it keeps the objects in an undesired state.

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

fromanirh commented on controllers/numaresourcesscheduler_controller.go:
```diff
@@ -109,8 +114,21 @@ func (r *NUMAResourcesSchedulerReconciler) Reconcile(ctx context.Context, req ct
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *NUMAResourcesSchedulerReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// we want to initiate reconcile loop only on change under labels or spec of the object
```
should we ignore annotations them? why or why not?

Tal-or replied:
I thought it might somehow reduce reconciliation loops, but maybe let's stay with the basics and reconcile for every change.

Tal-or replied:
I followed the controller code, there is no reason to act upon other changes because according to the code we won't revert metadata changes.
For ```Labels``` and ```Annotations``` changes, the controller will merge old and new values.
For changes under ```spec``` it will force the desired state.

fromanirh commented:
This seems like a good improvement, thanks for starting this work.

Tal-or commented:
I'm happy with the PR, lifting WIP

ffromani reviewed (commented):
/lgtm

fromanirh commented on test/e2e/sched/sched_test.go:
```diff
@@ -68,6 +72,88 @@ var _ = Describe("[Scheduler] imageReplacement", func() {
+			initialCM := nroCM.DeepCopy()
+			nroCM.Data["somekey"] = "somevalue"
+
+			err = e2eclient.Client.Update(context.TODO(), nroCM)
```
update can fail on a live cluster. We either need to use something smarter (patch?) or wrap the (get+)update in a loop

Tal-or replied:
Why that update would failed?

fromanirh replied:
I don't know the exact reason, but this is something we may happen any time in a cluster (part of how kube works) so robust code should tolerate it

Tal-or replied:
It's a bit worrying that a global wide-used function can fail without any good reason.
Does it mean we should stop using it? what about all the other places and production code we have which contain this function?

Of course, I can wrap it in a loop. But the general statement you raised here makes me question.

fromanirh replied:
Well this is kubernetes operates since day one pretty much. Retrying loop (which is what ultimately a control loop is about) and optimistic concurrency are the cornerstones of the platform. I'm not sure I understand the concern :\

ffromani reviewed (commented):
/lgtm cancel

Tal-or commented:
/test ci-e2e

fromanirh commented on controllers/numaresourcesscheduler_controller_test.go:
```diff
@@ -158,5 +158,52 @@ var _ = ginkgo.Describe("Test NUMAResourcesScheduler Reconcile", func() {
+			c := corev1.Container{Name: "newcontainer"}
+			dp.Spec.Template.Spec.Containers = append(dp.Spec.Template.Spec.Containers, c)
+
+			err = reconciler.Client.Update(context.TODO(), cm)
```
this has to be done in a `Eventually` loop

fromanirh commented on controllers/numaresourcesscheduler_controller_test.go:
```diff
@@ -158,5 +158,52 @@ var _ = ginkgo.Describe("Test NUMAResourcesScheduler Reconcile", func() {
+			gomega.Expect(err).ToNot(gomega.HaveOccurred())
+
+			key = client.ObjectKeyFromObject(cm)
+			err = reconciler.Client.Get(context.TODO(), key, cm)
```
probably same as above - I'm not sure why we need to retry a `Get` though

Tal-or replied:
I'm not sure it's necessary, there are plenty of examples in the code of ```Get``` but none of them is in ```Eventually``` loop except ones that are expected to be transitioned.

fromanirh commented on test/e2e/sched/sched_test.go:
```diff
@@ -68,6 +73,90 @@ var _ = Describe("[Scheduler] imageReplacement", func() {
+
+			nroSchedObj.Spec = objects.TestNROScheduler().Spec
+
+			err = e2eclient.Client.Update(context.TODO(), nroSchedObj)
```
let's do this inside a `Eventually` loop

fromanirh commented on test/e2e/sched/sched_test.go:
```diff
@@ -68,6 +73,90 @@ var _ = Describe("[Scheduler] imageReplacement", func() {
+			c := objects.NewTestPodPause("", "newcontainer").Spec.Containers[0]
+			dp.Spec.Template.Spec.Containers = append(dp.Spec.Template.Spec.Containers, c)
+
+			err = e2eclient.Client.Update(context.TODO(), dp)
```
should probably be done inside a `Eventually` loop

ffromani reviewed (commented):
we need to wrap updates inside `Eventually` loops, and we're good

Tal-or commented:
/retest

ffromani reviewed (commented):
/lgtm
thanks for the updates

Tal-or committed:
e2e: get error properly from `ForDeploymentComplete`

`ForDeploymentComplete` signature has changed.
The caller should change the way it handle the returned error

Signed-off-by: Talor Itzhak <titzhak@redhat.com>

fromanirh commented:
/lgtm
